src/main/kotlin/org/vitrivr/cottontail/database/index/pq/PQIndex.kt

```diff
@@ -0,0 +1,380 @@
+package org.vitrivr.cottontail.database.index.pq
+
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap
+import org.mapdb.DB
+import org.slf4j.LoggerFactory
+import org.vitrivr.cottontail.database.column.*
+import org.vitrivr.cottontail.database.entity.Entity
+import org.vitrivr.cottontail.database.entity.EntityTx
+import org.vitrivr.cottontail.database.events.DataChangeEvent
+import org.vitrivr.cottontail.database.index.Index
+import org.vitrivr.cottontail.database.index.IndexTx
+import org.vitrivr.cottontail.database.index.IndexType
+import org.vitrivr.cottontail.database.index.va.VAFIndex
+import org.vitrivr.cottontail.database.queries.components.AtomicBooleanPredicate
+import org.vitrivr.cottontail.database.queries.components.KnnPredicate
+import org.vitrivr.cottontail.database.queries.components.Predicate
+import org.vitrivr.cottontail.database.queries.planning.cost.Cost
+import org.vitrivr.cottontail.execution.TransactionContext
+import org.vitrivr.cottontail.math.knn.selection.ComparablePair
+import org.vitrivr.cottontail.math.knn.selection.MinHeapSelection
+import org.vitrivr.cottontail.math.knn.selection.MinSingleSelection
+import org.vitrivr.cottontail.model.basics.*
+import org.vitrivr.cottontail.model.exceptions.QueryException
+import org.vitrivr.cottontail.model.recordset.StandaloneRecord
+import org.vitrivr.cottontail.model.values.DoubleValue
+import org.vitrivr.cottontail.model.values.IntValue
+import org.vitrivr.cottontail.model.values.types.VectorValue
+import org.vitrivr.cottontail.utilities.extensions.write
+import org.vitrivr.cottontail.utilities.math.KnnUtilities
+import java.nio.file.Path
+import java.util.*
+import kotlin.collections.ArrayDeque
+
+/**
+ * An [Index] structure for nearest neighbor search (NNS) that uses a product quantization (PQ). Can
+ * be used for all type of [VectorValue]s and distance metrics.
+ *
+ * TODO: Check if generalization to other distance metrics is actually valid.
```

Comment:
I guess this needs quantification. In an ideal implementation, one would learn with the distance metric used during query time (i.e. L1 if you want to use L1 distance for querying). 

I wonder, however, how much of an impact this really has.

Reply:
Added issue #66 for this.

Reply:
> I guess this needs quantification, especially since the impact is probably highly dependent on the data. In an ideal implementation, one would probably learn with the distance metric used during query time (i.e. L1 if you want to use L1 distance for querying). I wonder, however, how much of an impact this really has in practice.

I agree.

> Would it be a solution to specify what kind of distance we intend to use and adjust learning based on this selection (i.e. make the distance function for the learning step configurable?)

I think that this would be the way to go for now. In my implementation I think I had a way to supply the covariance matrix. One could mimick standard k-means learning through this (identity matrix). This possibility should be there for another reason: in MIPS, the required cov-matrix is actually the *query* covariance matrix that could be estimated from a representative set of example queries and submitted at index-time. Would be great to have this to enable future refinements.